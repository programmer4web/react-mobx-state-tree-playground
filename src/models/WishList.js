import {types, getEnv, getParent, destroy } from 'mobx-state-tree';

import Category from './Category';

export const WishListItem = types.model({
  name: types.string,
  price: types.number,
  image: types.optional(types.string, ""), // or ""
  category: types.maybe(types.reference(Category))
})
.actions(self => ({
    changeName(newName) {
      self.name = newName;
    },
    changePrice(newPrice) {
      self.price = newPrice;
    },
    changeImage(newImage) {
      self.image = newImage;
    },
    changeCategory(newCategory) {
      self.category = newCategory;
    },
    remove() {
      getParent(self, 2).remove(self)
    }
  })
)


export const WishList = types.model({
    categories: types.array(Category),
    items: types.optional(types.array(WishListItem), [])
  }
)
.actions(self => ({
    add(item) {
      self.items.push(item);
    },
    remove(item) {
      destroy(item);
    },
    addCategory(category){
      const idx = self.categories.length;
      self.categories.push({
        id: idx.toString(),
        name: category.name
      })
    }
  })
).views(self => ({
  get totalPrice() {
    getEnv(this).alert('computed total price.');
    return self.items.reduce((sum, entry) => 
    { return sum + entry.price}, 0)
  }
}
))
