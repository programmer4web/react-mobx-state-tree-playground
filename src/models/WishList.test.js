import {WishListItem, WishList} from './WishList'
import {getSnapshot, onSnapshot, onPatch} from "mobx-state-tree";
import { reaction } from 'mobx';

it('can create an instance of the model', () => {
  const item = WishListItem.create({
    "name": "Cronicles of Narnia Box Set",
    "price": 28.73,
    "image": "https://images-na.ssl-images-amazon.com/images/I/51YGwxKn3CL._SX258_BO1,204,203,200_.jpg"
  })

  expect(item.price).toBe(28.73);
  item.changeName('Narnia');
  expect(item.name).toBe('Narnia');

})

// it('can create an wishlist', () => {
//   const states = [];
//   const list = WishList.create({
//     items: [{
//     "name": "Cronicles of Narnia Box Set",
//     "price": 28.73,
//     "image": "https://images-na.ssl-images-amazon.com/images/I/51YGwxKn3CL._SX258_BO1,204,203,200_.jpg"
//     }]
//   })

//   onSnapshot(list, snapshot => {
//     states.push(list)
//   })

//   expect(list.items.length).toEqual(1);
//   expect(getSnapshot(list)).toMatchSnapshot();
//   expect(states).toMatchSnapshot();
// })

it('can add new items', ()=> {
  const list = WishList.create();
  list.add(
    WishListItem.create({
      name: 'Chesterton',
      price: 12
    })
  )

  expect(list.items.length).toBe(1);
  list.items[0].changeImage('test');
  expect(list.items[0].image).toBe('test');
})

it('can add new items 2', () => {
  const list = WishList.create(),
  patches = [];
  list.add(
    WishListItem.create({
      name: 'Chesterton',
      price: 12
    })
  )


  onPatch(list, patch => {
    patches.push(list)
  })

 list.items[0].changeName("The book with new name");
expect(patches).toMatchSnapshot();
})


it('can calculate total Price', () => {
  const list = WishList.create();
  list.add(
    WishListItem.create({
      name: 'Chesterton',
      price: 12
    })
  )
  list.add(
    {
      name: 'Sana',
      price: 3
    }
  )

  expect(list.totalPrice).toBe(15);
  let changed = 0;
  reaction(()=> list.totalPrice, () => changed++)
  console.log(changed);
  list.items[0].changeName('Gefanada');
  console.log(changed);
})