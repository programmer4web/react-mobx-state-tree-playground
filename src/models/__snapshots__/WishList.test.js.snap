// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`can add new items 2 1`] = `
Array [
  Object {
    "categories": Array [],
    "items": Array [
      Object {
        "category": undefined,
        "image": "",
        "name": "The book with new name",
        "price": 12,
      },
    ],
  },
]
`;
